cmd/skopeo/proxy.go

```diff
@@ -300,15 +297,82 @@ func (h *proxyHandler) allocPipe() (*os.File, *activePipe, error) {
 	return piper, &f, nil
 }
 
+// returnBytes generates a return pipe() from a byte array
+// In the future it might be nicer to return this via memfd_create()
+func (h *proxyHandler) returnBytes(retval interface{}, buf []byte) (replyBuf, error) {
+	var ret replyBuf
+	piper, f, err := h.allocPipe()
+	if err != nil {
+		return ret, err
+	}
+
+	go func() {
+		// Signal completion when we return
+		defer f.wg.Done()
+		_, err = io.Copy(f.w, bytes.NewReader(buf))
+		if err != nil {
+			f.err = err
+		}
+	}()
+
+	ret.value = retval
+	ret.fd = piper
+	ret.pipeid = uint32(f.w.Fd())
+	return ret, nil
+}
+
+// cacheTargetManifest is invoked when GetManifest or GetConfig is invoked
+// the first time for a given image.  If the requested image is a manifest
+// list, this function resolves it to the image matching the calling process'
+// operating system and architecture.
+//
+// TODO: Add GetRawManifest or so that exposes manifest lists
```

Comment:
> Alternatively, it might make sense to actually design for multi-arch, in some way, at least enough to see what the future would look like. ...
>  It just feels like a kind of thing where a little thought advance could save some pain in the future.

Yeah. I don't have an immediate need to support non-current arch images but it definitely makes sense to consider it.

> For context, the c/image approach which completely silently converts multi-arch manifest lists into current-arch images, with the weird warts like still returning the raw multi-arch manifest list, was created in a sort of panic after Docker Hub started serving clients multi-arch overnight. It’s not necessarily a well-thought-out model to follow — but then we never had the option to consider completely ripping that approach out, so I also don’t have a strong opinion on what would be a good model.)

Ahh, I see.  That historical context helps explain things.